### PR TITLE
Cache datasets when they're downloaded.

### LINKED ISSUE #375: Hardcoded invalid dataset location in the npm dist
It seems like your build script for dist package is messing with paths.

"propel@3.2.0" from npm contains lots of lines like this one:

`var __dirname = "/Users/rld/src/propel/src";`

Which causes errors when loading datasets.

### PR DESCRIPTION
**Pull request by ry**
Fixes #375

### CODE REVIEW

**Comment by ry** [comment]
@piscisaureus PTAL

File: src/cache.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// The $HOME/.propel/cache directory is where these files will be stored.
+
+import * as path from "path";
+import * as rimraf from "rimraf";
+import { assert, fetchArrayBuffer, IS_WEB, nodeRequire, URL } from "./util";
+import { mkdirp, propelDir } from "./util_node";
+
+export interface Cache {
+  clearAll(): Promise<void>;
+  get(url: string): Promise<null | ArrayBuffer>;
```
[the remark this answers is not in the document]
Comment (ry) [reply src/cache.ts]: Changed.

File: src/cache.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// "$HOME/.propel/cache/propelml.org/data/mnist/train-images-idx3-ubyte.bin"
+export function url2Filename(url: string): string {
+  // Throw on browser. We expose this method for testing, but only run it on
+  // Node.
+  assert(!IS_WEB, "url2Filename is unsupposed in the browser");
+  const u = new URL(url);
+  if (!(u.protocol === "http:" || u.protocol === "https:")) {
+    throw Error(`Unsupported protocol '${u.protocol}'`);
+  }
+  if (u.pathname.indexOf("..") >= 0) {
```
[the remark this answers is not in the document]
Comment (ry) [reply src/cache.ts]: Done.

File: src/cache.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  // Note we purposely leave the port out of the cache path because
+  // Windows doesn't allow colons in filenames. This is probably fine
+  // in 99% of cases and is the simplest solution.
+  const cacheFn = path.resolve(path.join(cacheBase(), u.hostname, u.pathname));
+  assert(cacheFn.startsWith(cacheBase()));
+  return cacheFn;
+}
+
+if (IS_WEB) {
+  // On web do nothing. No caching.
```
[the remark this answers is not in the document]
Comment (ry) [reply src/cache.ts]: Noted.

File: src/cache.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  assert(!IS_WEB, "url2Filename is unsupposed in the browser");
+  const u = new URL(url);
+  if (!(u.protocol === "http:" || u.protocol === "https:")) {
+    throw Error(`Unsupported protocol '${u.protocol}'`);
+  }
+  if (u.pathname.indexOf("..") >= 0) {
+    throw Error("Cache name cannot include '..'");
+  }
+  // Note we purposely leave the port out of the cache path because
+  // Windows doesn't allow colons in filenames. This is probably fine
```
[the remark this answers is not in the document]
Comment (ry) [reply src/cache.ts]: I'm happy to leave out the port until we hit a bug - which will likely be never.

File: src/cache.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+// a training program, we provide a local cache of these datasets.
+// The $HOME/.propel/cache directory is where these files will be stored.
+
+import * as path from "path";
+import * as rimraf from "rimraf";
+import { assert, fetchArrayBuffer, IS_WEB, nodeRequire, URL } from "./util";
+import { mkdirp, propelDir } from "./util_node";
+
+export interface Cache {
+  clearAll(): Promise<void>;
```
[the remark this answers is not in the document]
Comment (ry) [reply src/cache.ts]: I'm not trying to implement the map interface. If anything it should implement this interface https://developer.mozilla.org/en-US/docs/Web/API/Cache 
But I think it's fine for now.

File: src/cache.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/cache.ts]: ok

File: src/cache.ts [hunk 6]
```diff
@@ -61,13 +61,15 @@ export function url2Filename(url: string): string {
   if (!(u.protocol === "http:" || u.protocol === "https:")) {
     throw Error(`Unsupported protocol '${u.protocol}'`);
   }
-  if (u.pathname.indexOf("..") >= 0) {
-    throw Error("Cache name cannot include '..'");
-  }
+  // Just to be safe, use encodeURI to transform pathname and hostname.
+  const h = encodeURI(u.hostname);
```
Comment (piscisaureus) [inline src/cache.ts]: I think it should be encodeURIComponent() and not encodeURI().
But then I was trying this: `encodeURIComponent("*:<>|..")`, and the result still isn't a valid file name on Windows. But then on the other hand, I've never seen an URL like that either, so maybe this is too academic and you should just add a comment that there is a loose end here.
One day somebody will be affected by this though, they'll get an ENOENT error when attempting to write a cache file.

Comment (ry) [reply src/cache.ts]: Added to the comment.

**Pull request merged**